packages/devtools/src/utils.ts

```diff
@@ -30,3 +42,140 @@ export async function promiseMap<
     return acc
   }, {} as Record<keyof M, R>)
 }
+
+const ceramicTypeDefs = `
+# From graphql-scalars
+
+scalar CountryCode
+scalar Date
+scalar DateTime
+scalar DID
+scalar PositiveInt
+scalar Time
+scalar URL
+
+# Custom Ceramic scalars
+
+scalar StreamID
+
+# Field validation directives
+
+directive @ipfs on FIELD_DEFINITION # Must be an IPFS URL
+directive @length(max: Int!, min: Int = 0) on FIELD_DEFINITION
+
+# Metadata access directives
+
+directive @documentAccount on FIELD_DEFINITION
+directive @documentVersion on FIELD_DEFINITION
+
+# Model definition
+
+enum ModelIndexType {
+  LIST # Account to multiple streams - default
+  SET # Account to multiple streams but only one reference
+  LINK # Account to single stream (IDX)
+  NONE # Indexing explicitly disabled
+}
+directive @model(
+  index: ModelIndexType = LIST
+  description: String
+  url: String
+) on OBJECT
+# When a model has index = SET, at least one field must have the @index directive
+directive @index on FIELD_DEFINITION
+
+# Account-based relations
+
+enum AccountLinkTarget {
+  SELF
+  OTHER
+  BOTH
+}
+directive @accountLink(
+  property: String!
+  target: AccountLinkTarget!
+) on FIELD_DEFINITION
+directive @accountReference on FIELD_DEFINITION
+directive @accountRelation(property: String!) on FIELD_DEFINITION
+
+# Model-based relations
+
+directive @modelReference(type: String!) on FIELD_DEFINITION
+directive @modelRelation(property: String!) on FIELD_DEFINITION
+`
+
+/** @internal */
+export function buildCompositeSchema(sdl: string): GraphQLSchema {
+  return (
+    lengthDirectiveTransformer(
+      modelDirectiveTransformer(
+        ipfsDirectiveTransformer(
+          makeExecutableSchema({typeDefs: [ceramicTypeDefs, sdl]})
+        )
+      )
+    )
+  ) 
+}
+
+function fieldSchemaFromFieldDefinition(
+  fieldDefinition: Record<string, any>
+  ): Record<string, any> | undefined {
+  const ceramicExtensions = fieldDefinition.extensions?.ceramicExtensions
+  if (ceramicExtensions?.length !== undefined) {
+    return {
+      type: 'string',
+      title: fieldDefinition.name,
+      maxLength: ceramicExtensions.length.max,
+      minLength: ceramicExtensions.length.min,
+    }
+  } else if (ceramicExtensions?.ipfs !== undefined) {
+    return {
+      type: 'ipfsUrlString',
```

Comment:
`type` should be a valid JSON schema type, I don't think we can create custom types like this.